processing/src/main/java/io/druid/query/extraction/FunctionalExtraction.java

```diff
@@ -0,0 +1,118 @@
+
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ * Functional extraction uses a function to find the new value.
+ * null values in the range can either retain the domain's value, or replace the null value with "replaceMissingValueWith"
+ */
+public abstract class FunctionalExtraction extends DimExtractionFn
+{
+  private final boolean retainMissingValue;
+  private final String replaceMissingValueWith;
+  private final Function<String, String> extractionFunction;
+  private final ExtractionType extractionType;
+
+  /**
+   * The general constructor which handles most of the logic for extractions which can be expressed as a function of string-->string
+   *
+   * @param extractionFunction      The function to call when doing the extraction. The function must be able to accept a null input.
+   * @param retainMissingValue      Boolean value on if functions which result in `null` should use the original value or should be kept as `null`
+   * @param replaceMissingValueWith String value to replace missing values with (instead of `null`)
+   * @param injective               If this function always has 1:1 renames and the domain has the same cardinality of the input, this should be true and enables optimized query paths.
+   */
+  public FunctionalExtraction(
+      final Function<String, String> extractionFunction,
+      final boolean retainMissingValue,
+      final String replaceMissingValueWith,
+      final boolean injective
+  )
+  {
+    this.retainMissingValue = retainMissingValue;
+    this.replaceMissingValueWith = Strings.isNullOrEmpty(replaceMissingValueWith) ? null : replaceMissingValueWith;
+    Preconditions.checkArgument(
+        !(this.retainMissingValue && !Strings.isNullOrEmpty(this.replaceMissingValueWith)),
+        "Cannot specify a [replaceMissingValueWith] and set [retainMissingValue] to true"
+    );
+
+    // If missing values are to be retained, we have a slightly different code path
+    // This is intended to have the absolutely fastest code path possible and not have any extra logic in the function
+    if (this.retainMissingValue) {
+
+      this.extractionFunction = new Function<String, String>()
+      {
+        @Nullable
+        @Override
+        public String apply(@Nullable String dimValue)
+        {
+          final String retval = extractionFunction.apply(dimValue);
+          return Strings.isNullOrEmpty(retval) ? (Strings.isNullOrEmpty(dimValue) ? null : dimValue) : retval;
+        }
+      };
+    } else {
+      this.extractionFunction = new Function<String, String>()
+      {
+        @Nullable
+        @Override
+        public String apply(@Nullable String dimValue)
+        {
+          final String retval = extractionFunction.apply(dimValue);
+          return Strings.isNullOrEmpty(retval) ? FunctionalExtraction.this.replaceMissingValueWith : retval;
```

Comment:
And if the class reference is removed completely it tries to pick up the method parameter, which is not the same as the final class field.